synapse/handlers/profile.py

```diff
@@ -157,6 +157,13 @@ def set_displayname(self, target_user, requester, new_displayname, by_admin=Fals
         if not by_admin and target_user != requester.user:
             raise AuthError(400, "Cannot set another user's displayname")
 
+        if not by_admin and self.hs.config.disable_set_displayname:
+            profile = yield self.store.get_profileinfo(target_user.localpart)
+            if profile.display_name:
+                raise SynapseError(
+                    400, "Changing displayname is disabled on this server"
```

Comment:
This error will probably be shown to the user by the client, therefore

```suggestion
                    400, "Changing display name is disabled on this server"
```

might be better as it makes it easier to understand what we're talking about (imho).

You also probably want to use a custom error code, according to https://matrix.org/docs/spec/client_server/latest#api-standards `M_FORBIDDEN` sounds like the most relevant in this case. FWIW there are constants for the error codes in `synapse.api.errors` (in the class `Codes`, e.g. `Codes.FORBIDDEN`). Same goes with all the `raise SynapseError` this PR adds IMO (since they're all about forbidding the user from doing something).